### PR TITLE
[quantization] Support PTQ quantization for VLM Qwen3-vl

### PR DESCRIPTION
**Pull request by Torrero**
This PR introduces PTQ quantization for VLM Qwen3-vl.

For benchmarks evaluation `Qwen3VLForConditionalGeneration` has been inherited from the `GenerationMixin`.
In addition, `quant_vision_model.py` was changed by introducing logic for dynamically computing `position_embeddings`,`cu_seqlens`, or using precomputed values with a fixed `grid_thw` to export a static graph.

TICO-DCO-1.0-Signed-off-by:  Evgenii Maltsev <e.maltsev@samsung.com>

### CODE REVIEW

File: tico/quantization/wrapq/wrappers/qwen_vl/quant_vision_model.py [hunk 1]
```diff
@@ -383,20 +383,37 @@ def forward(
...
+        #vision_output = self.module(hidden_states, grid_thw=grid_thw, **kwargs)
+        #return vision_output
+
+        # Model export mode (torch.export.export) requires the use of fixed precomputed values (`grid_thw`,`position_embeddings`,`cu_seqlens').
+        # `torch.compiler.is_compiling()` controls the conditional behavior of the model:
+        # - precomputed values are used in export mode
+        # - otherwise, the calculation is performed dynamically(e.g. benchmarks evaluation)
+        if torch.compiler.is_compiling():
+            # Assert that grid_thw matches the precomputed vision_grid_thw
+            if self._mode is Mode.CALIB:
```
Comment (dvsav) [inline tico/quantization/wrapq/wrappers/qwen_vl/quant_vision_model.py]: I guess, when we are converting a model to Circle, it's in QUANT mode, isn't it? So, maybe asserting in QUANT mode would be more logical.

Comment (Torrero) [reply tico/quantization/wrapq/wrappers/qwen_vl/quant_vision_model.py]: Yes, you are right, thank you!

**Comment by Torrero** [comment]
## `PTQ` for  `Qwen3-vl`
There is an issue with outliers in `LayerNorm` operation in [norm1](https://github.com/Samsung/TICO/blob/70867ffc705ae198bea307dd50ac7277e487bd05/tico/quantization/wrapq/wrappers/qwen_vl/quant_vision_block.py#L118), [norm2](https://github.com/Samsung/TICO/blob/70867ffc705ae198bea307dd50ac7277e487bd05/tico/quantization/wrapq/wrappers/qwen_vl/quant_vision_block.py#L137) of `quant_vision_block` wrapper, and [norm](https://github.com/Samsung/TICO/blob/70867ffc705ae198bea307dd50ac7277e487bd05/tico/quantization/wrapq/wrappers/qwen_vl/quant_vision_patch_merger.py#L99) in `quant_vision_patch_merger` wrapper. 

### The reason is [square](https://github.com/Samsung/TICO/blob/70867ffc705ae198bea307dd50ac7277e487bd05/tico/quantization/wrapq/wrappers/nn/quant_layernorm.py#L121) 
image  in `LayerNorm` operation that leads to wide range of `Min Max` values and a rough `scale`. 


════════════════════════════════════════
  tensor distribution of `square` operation in  [norm1](https://github.com/Samsung/TICO/blob/70867ffc705ae198bea307dd50ac7277e487bd05/tico/quantization/wrapq/wrappers/qwen_vl/quant_vision_block.py#L118) operation of original model
════════════════════════════════════════
  Shape: (1200, 1024) | Elements: 1,228,800
  Min: 0.0000 | Max: 60832.9688
  Mean: 0.4465 | Std: 113.2365 | Median: 0.0228
```
                                  Value Distribution                            
       ┌───────────────────────────────────────────────────────────────────────┐
1228776┤•                                                                      │
       │                                                                       │
1023980┤                                                                       │
       │                                                                       │
 819184┤                                                                       │
       │                                                                       │
 614388┤                                                                       │
       │                                                                       │
       │                                                                       │
 409592┤                                                                       │
       │                                                                       │
 204796┤                                                                       │
       │                                                                       │
      0┤  •  • •  • • •  • •  • •  • • •  • •  • • •  • •  • •  • • •  • •  • •│
       └┬─────────────────┬────────────────┬─────────────────┬────────────────┬┘
     1013.9            15715.2          30416.5           45117.8       59819.1 
Count                                    Value   
```

`QScheme.PER_TENSOR_SYMM` for the `square` operation significantly degrades the quantization accuracy even for W16A16  (`vqav2` dataset): **0.85** (original) -> **0.44** (quantized W16A16).
Applying `QScheme.PER_CHANNEL_ASYMM` with `EMAObserver` for [square](https://github.com/Samsung/TICO/blob/70867ffc705ae198bea307dd50ac7277e487bd05/tico/quantization/wrapq/wrappers/nn/quant_layernorm.py#L121) operation produces better results (**0.84**).
But I'm not sure about applying the `PER_CHANNEL_ASYMM` QScheme for intermediate activations. 
@mhs4670go, could you please provide your opinion about application of `PER_CHANNEL_ASYMM` QScheme for `square` operation. 


Outdated:
<details>
Preliminary test with 100 nsamples:
### **qwen3-vl-2B**
nsamples_for_evaluation: 100
nsamples_for_qcalibration: 128
Calib seq len    : 2048
Max seq len      : 2048

|model|vqav2|
|--|--|
|original|0.85|
|W16A16,  `square` op QScheme  - `PER_TENSOR_SYMM`  | 0.44 |
|W16A16, `square` op QScheme  - `PER_CHANNEL_ASYMM`  | 0.84 |
|W4A16 + GPTQ + MSE, `square` op QScheme - `PER_CHANNEL_ASYMM`  | 0.82 |

</details>

**Comment by Torrero** [comment]
I see, export to `circle` does not support this type of quantization for activations. 😭

File: tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py
Comment (dvsav) [inline tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py:613]: Would you consider extraction of this stuff to a separate `pargse_args()` function?

File: tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py [hunk 2]
```diff
@@ -195,9 +202,121 @@ def print_markdown_comparison(
...
+        lm_head_weight_bits=args.lm_head_weight_bits,
+        norm_dtype=DType.int(16),
+        norm_weight_dtype=DType.int(16),
+        quantize_vision=args.quantize_vision,
+        quantize_text=args.quantize_text,
+        quantize_lm_head=args.quantize_lm_head,
+        strict_wrap=True,
+    )
+
+    q_m = prepare(q_m, qcfg)
```
Comment (dvsav) [inline tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py:290]: `quant_model` or `quantized_model` might be more readable.

File: tico/quantization/wrapq/wrappers/qwen_vl/quant_for_conditional_generation.py [hunk 3]
```diff
@@ -38,6 +40,10 @@ class QuantQwen3VLForConditionalGeneration(QuantModuleBase):
     with no additional quantization operations needed at this level.
     """
 
+    main_input_name = "input_ids"
+    _is_stateful = True
+    _supports_cache_class = True
```
Comment (dvsav) [inline tico/quantization/wrapq/wrappers/qwen_vl/quant_for_conditional_generation.py:45]: ### Note for me

`_is_stateful` and `_supports_cache_class` are internal Boolean flags used within the [Hugging Face Transformers library](https://www.google.com/url?sa=i&source=web&rct=j&url=https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_utils.py&ved=2ahUKEwi69cbrvIaUAxXAUkEAHVviNi8Qy_kOegQIARAB&opi=89978449&cd&psig=AOvVaw22uSe5CLrEwkKbBfYOKac7&ust=1777119680600000) to manage model behavior during generation. They define if a model maintains state (like State Space Models - SSMs) or supports KV caching (DynamicCache/StaticCache). 

- `_is_stateful`: Indicates if the model maintains an internal state that needs to be updated during generation.
- `_supports_cache_class`: Indicates if the model supports specific KV cache classes (cache_class) used to speed up generation by storing past token keys and values.

File: tico/quantization/wrapq/wrappers/qwen_vl/quant_vision_model.py [hunk 4]
```diff
@@ -405,19 +421,32 @@ def forward(
...
+        if torch.compiler.is_compiling():
+            # Use precomputed RoPE position embeddings (cos, sin) and quantize
+            cos = self.rope_cos_template.to(
+                dtype=hidden_states.dtype, device=hidden_states.device
+            )
+            sin = self.rope_sin_template.to(
+                dtype=hidden_states.dtype, device=hidden_states.device
+            )
+        else:
+            inv_freq = self._precompute_rope_inv_freq(dim=self.dim, theta=self.theta)
```
Comment (dvsav) [inline tico/quantization/wrapq/wrappers/qwen_vl/quant_vision_model.py]: Can we reuse `rope_inv_freq`?
It looks like the computations are identical:
```python
def __init__(...):
        inv_freq = self._precompute_rope_inv_freq(dim=self.dim, theta=self.theta)
        self.register_buffer("rope_inv_freq", inv_freq, persistent=False)
```

Comment (Torrero) [reply tico/quantization/wrapq/wrappers/qwen_vl/quant_vision_model.py]: Changed it, thank you

File: test/quantization/wrapq/wrappers/qwen_vl/test_quant_vision_model.py [hunk 5]
```diff
@@ -245,21 +245,6 @@ def test_mode_transitions(self):
...
-        )
-        q_model.enable_calibration()
-
-        # Try with different grid
-        hidden_states, grid_thw = self._create_test_inputs((1, 4, 4))
-
-        with self.assertRaises(AssertionError) as context:
-            _ = q_model(hidden_states, grid_thw)
-        self.assertIn("grid_thw", str(context.exception))
-
```
Comment (Torrero) [inline test/quantization/wrapq/wrappers/qwen_vl/test_quant_vision_model.py:262]: This test is no longer needed, as the `grid_thw` mismatch assert in `quant_vision_model` has been eliminated.

File: tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py (hunk 2 above)
Comment (Torrero) [reply tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py:290]: I tried to make fewer changes in this PR for a simpler review, I think we can apply this refactoring in further PRs

**Comment by Torrero** [comment]
Preliminary test for 
### **qwen3-vl-4B**
nsamples_for_evaluation  : 1000
nsamples_for_qcalibration : 128
Calib seq len                      : 2048
Max seq len                       : 2048


linear_weight_bits              : 4 
vision_patch_embed_weight_bits : 4 
embedding_weight_bits    : 8
lm_head_weight_bits         : 4

|model|vqav2|
|--|--|
|original|0.895|
|W4A16 + GPTQ + MSE | 0.691 (-20%)|
|W4A16 + GPTQ + SMSE | 0.695 (-20%)|

Additional algorithms should be applied for improving accuracy degradation.

Run command:
```
python tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py --model=Qwen/Qwen3-VL-4B-Instruct --trust-remote-code --eval_tasks=vqav2 --gptq_mse=smse --nsamples_for_evaluation=1000 --embedding_weight_bits=8 --vision_patch_embed_weight_bits=4 --linear_weight_bits=4 --lm_head_weight_bits=4 --nsamples_for_qcalibration=128 
```

File: tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py
Comment (Torrero) [reply tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py:613]: This is a good point, thank you, let's apply it in the next PR.

File: test/quantization/wrapq/wrappers/qwen_vl/test_quant_vision_model.py (hunk 5 above)
Comment (dvsav) [reply test/quantization/wrapq/wrappers/qwen_vl/test_quant_vision_model.py:262]: **Note for me**: assertion of equality of `grid_thw` argument and `vision_grid_thw` configuration parameter hindered quantized model evaluation with standard datasets. On the other hand, making that assertion at export time is not possible either because assert operator is not exportable.

File: tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py [hunk 6]
```diff
@@ -491,6 +673,64 @@ def main() -> None:
...
+            else:
+                raise ValueError(
+                    "Cannot determine num_vision_blocks from vision_config. "
+                    "Expected vision_config.num_hidden_layers, num_layers, or depth."
+                )
+        else:
+            raise ValueError(
+                "Cannot determine num_vision_blocks from model config. "
+                "Please ensure the model has config.vision_config."
+            )
```
Comment (dvsav) [inline tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py]: (Optional)
I'd suggest encapsulating this code in a function, e.g. `num_vision_blocks = get_num_vision_blocks(q_m)`.

File: tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py [hunk 7]
```diff
@@ -491,6 +673,64 @@ def main() -> None:
...
+        if hasattr(q_m, "config") and hasattr(q_m.config, "text_config"):
+            num_text_layers = q_m.config.text_config.num_hidden_layers
+        elif hasattr(q_m, "config") and hasattr(q_m.config, "num_hidden_layers"):
+            num_text_layers = q_m.config.num_hidden_layers
+        else:
+            raise ValueError(
+                "Cannot determine num_text_layers from model config. "
+                "Please ensure the model has config.text_config.num_hidden_layers "
+                "or config.num_hidden_layers."
+            )
```
Comment (dvsav) [inline tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py]: (Optional)
I'd suggest encapsulating this code in a function, e.g. `num_text_layers = get_num_text_layers(q_m)`.

File: tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py [hunk 8]
```diff
@@ -491,6 +673,64 @@ def main() -> None:
...
+
+        # Get number of deepstack mergers from model config
+        num_deepstack_mergers = 0
+        if hasattr(q_m, "config") and hasattr(q_m.config, "vision_config"):
+            vision_config = q_m.config.vision_config
+            if hasattr(vision_config, "deepstack_visual_indexes"):
+                num_deepstack_mergers = len(vision_config.deepstack_visual_indexes)
```
Comment (dvsav) [inline tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py]: (Optional)
I'd suggest encapsulating this code in a function, e.g. `num_deepstack_mergers = get_num_deepstack_mergers (q_m)`.

**Review by dvsav — DISMISSED** [review]
LGTM

File: tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py (hunk 6 above)
Comment (Torrero) [reply tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py]: fixed this and the following comments, thanks

**Review by dvsav — APPROVED** [review]

**Review by mhs4670go — APPROVED** [review]
LGTM. Thank you!!
